### PR TITLE
Migrate to python 3

### PR DESCRIPTION
**Pull request by DJStompZone**
Was not working in any sense of the word on python 3.10, fixed at least a dozen bugs and restored to working order. Though I'm sure there's things I will have missed, but it runs from start to finish now, so that's progress!

### CODE REVIEW

**Comment by SOF3** [comment]
the commit title is not accurate. these are not "bugs". they are completely fine with python2.

**Comment by DJStompZone** [comment]
Cool, if I meet any time travelers from 2008 when that was relevant they will be delighted to hear it. In the meantime, feel free to use the version I fixed to work on versions of python under the age of 13. ✌️

**Comment by SOF3** [comment]
What I'm saying is that the purpose of this PR is to migrate to python3, not to fix bugs, because they are not bugs at all if you used the correct tools to run it. It is like trying to run a bash script with php and claiming that the bash script doesn't work because php doesn't interpret it properly.

File: protocol_info_generator_objdump.py [hunk 1]
```diff
@@ -13,10 +13,17 @@
 def convert_windows_path(path):
     return path.replace('\\', '/').replace('C:', '/mnt/c')
 
-bds_path = convert_windows_path(sys.argv[1])
-bedrockprotocol_path = convert_windows_path(sys.argv[2])
-print 'Dumping data from ' + bds_path
-print 'BedrockProtocol path ' + bedrockprotocol_path
+def valid_paths(list_of_paths):
+    results = [os.path.exists(each) for each in list_of_paths]
+    return results == [True]*len(list_of_paths)
```
Comment (SOF3) [inline protocol_info_generator_objdump.py]: simplify this to `all(os.path.exists(path) for path in list_of_paths)`?

**Review by SOF3 — CHANGES_REQUESTED** [review]

**Comment by SOF3** [comment]
to be clear, I am personally totally fine with this PR. I am just pointing out that python2 and python3 are basically two different languages, so it is not technically a "bug" in this sense, but it is still totally reasonable to update to python3 if someone wants to maintain it.

**Review by SOF3 — APPROVED** [review]

File: protocol_info_generator.py [hunk 2]
```diff
@@ -19,16 +18,17 @@ def generate_new_packet_stubs(out, resource_path, packets_dir):
...
             base_name = rchop(i, 'Packet')
             with open(packets_dir + i + '.php', 'w') as out_file:
                 out_file.write(packet_template % (i, const_name, base_name))
 
-            print 'Created stub class for', i, 'at', (packets_dir + i + '.php')
+            print('Created stub class for', i, 'at', (packets_dir + i + '.php'))
 
 def check_removed_packets(out, packets_dir):
     names = out.values()
+    print(f"Collected {len(out.values())} packet definitions\n")
```
Comment (dktapps) [inline protocol_info_generator.py:31]: This change appears unrelated to the topic of the PR

File: protocol_info_generator_objdump.py [hunk 3]
```diff
@@ -142,9 +148,10 @@ def dump_version():
...
             protocol = get_value_at(bds_path, int('0x' + start, 16) + rodata_shift, int('0x' + size, 16), 'i')
 
-    print major, minor, patch, revision, beta, protocol
+    print(major, minor, patch, revision, beta, protocol)
     return Version(major, minor, patch, revision, beta, protocol)
 
-version = dump_version()
-packets = dump_packet_ids()
-generate_stuff(packets, version, bedrockprotocol_path)
+if __name__ == "__main__":
```
Comment (dktapps) [inline protocol_info_generator_objdump.py:154]: This change doesn't seem related to the topic of the PR

File: protocol_info_generator_objdump.py [hunk 4]
```diff
@@ -13,10 +13,16 @@
...
-print 'Dumping data from ' + bds_path
-print 'BedrockProtocol path ' + bedrockprotocol_path
+def valid_paths(list_of_paths):
+    return all(os.path.exists(path) for path in list_of_paths)
+
+
+bds_path = convert_windows_path(sys.argv[1]) if os.name == 'nt' else sys.argv[1]
+bedrockprotocol_path = convert_windows_path(sys.argv[2]) if os.name == 'nt' else sys.argv[2] 
+if not valid_paths([bds_path, bedrockprotocol_path]):
+    raise Exception(f"Invalid arguments provided:{sys.argv[1:3]} check that the paths are correct")
```
Comment (dktapps) [inline protocol_info_generator_objdump.py:23]: This change is not related to the PR topic

File: protocol_info_generator.py [hunk 5]
```diff
@@ -114,15 +115,18 @@ def generate_packet_handler_interface(out, resource_path, packets_dir):
...
-        out_file.write(packet_handler_template % ('\n\n'.join(functions)))
+        out_file.write(packet_handler_template % (bytes('\n\n'.join(functions), "utf8")))
 
-    print 'Recreated packet handler interface'
+    print('Recreated packet handler interface')
 
 def generate_stuff(out, version, bedrockprotocol_dir):
+    timestamp = "-".join([str(ea) for ea in time.localtime()[3:6]])
+    with open(f"log-{timestamp}.json", "w") as fp:
+        fp.write(json.dumps({f"v{str(version).split(' ')[-1]}": out}))
```
Comment (dktapps) [inline protocol_info_generator.py:125]: This change is not related to the PR topic

**Review by dktapps — CHANGES_REQUESTED** [review]
This PR contains multiple changes that are not related in any way to a python3 migration. I'm not against most of them, but they don't belong in this PR and should be done in separate PRs.

**Comment by DJStompZone** [comment]
> This change is not related to the PR topic

That's why the original topic was "numerous bug fixes", as these were simply the changes required to successfully execute BedrockProtocolDumper on my machine with a modern version of python.
Seems like a silly and pedantic thing to quibble about though, really. It could be called "double cheeseburger and medium fry", and the code would still run exactly the same, so... Who cares what the title is?

**Comment by dktapps** [comment]
The title isn't really the point. Generally speaking it's bad PR or commit hygiene to put multiple unrelated changes into the same PR or commit.

**Comment by SOF3** [comment]
as a quick script that was not in a well maintained state, usually "code cleanup" is not efficient and understandable than atomic commits, especially when you most likely won't want to roll back small parts of the commit.

**Comment by dktapps** [comment]
This PR (and repository) has been superseded by the improved version of the tool in [bds-modding-devkit](https://github.com/pmmp/bds-modding-devkit) which uses Python 3 and generates JSON data instead of PHP code.

**Pull request closed without merging**